File: mock-server/src/main/java/com/buyansong/protocol/ImPacketCodec.java
package com.buyansong.imserver.protocol;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.zip.CRC32;

public final class ImPacketCodec {
    public static final short MAGIC = (short) 0xCAFE;
    public static final byte VERSION = 1;

    private static final int HEADER_SIZE = 11;
    private static final int CRC_SIZE = 4;

    private ImPacketCodec() {
    }

    public static byte[] encode(ImPacket packet) {
        byte[] body = packet.body();
        ByteBuffer buffer = ByteBuffer
                .allocate(HEADER_SIZE + body.length + CRC_SIZE)
                .order(ByteOrder.BIG_ENDIAN);
        buffer.putShort(MAGIC);
        buffer.put(VERSION);
        buffer.putInt(body.length);
        buffer.putInt(packet.cmd());
        buffer.put(body);
        int crcOffset = buffer.position();
        buffer.putInt(crc32(buffer.array(), 0, crcOffset));
        return buffer.array();
    }

    public static ImPacket decode(byte[] bytes) {
        if (bytes.length < HEADER_SIZE + CRC_SIZE) {
            throw new ProtocolException("Packet too short");
        }
        ByteBuffer crcBuffer = ByteBuffer.wrap(bytes, bytes.length - CRC_SIZE, CRC_SIZE).order(ByteOrder.BIG_ENDIAN);
        int expectedCrc = crcBuffer.getInt();
        int actualCrc = crc32(bytes, 0, bytes.length - CRC_SIZE);
        if (expectedCrc != actualCrc) {
            throw new ProtocolException("Invalid CRC");
        }

        ByteBuffer buffer = ByteBuffer.wrap(bytes).order(ByteOrder.BIG_ENDIAN);
        short magic = buffer.getShort();
        if (magic != MAGIC) {
            throw new ProtocolException("Invalid magic");
        }
        byte version = buffer.get();
        if (version != VERSION) {
            throw new ProtocolException("Unsupported version");
        }
        int length = buffer.getInt();
        if (length < 0 || bytes.length != HEADER_SIZE + length + CRC_SIZE) {
            throw new ProtocolException("Invalid body length");
        }
        int cmd = buffer.getInt();
        byte[] body = new byte[length];
        buffer.get(body);
        return new ImPacket(cmd, body);
    }

    private static int crc32(byte[] bytes, int offset, int length) {
        CRC32 crc = new CRC32();
        crc.update(bytes, offset, length);
        return (int) crc.getValue();
    }
}

File: app/src/main/java/com/buyansong/im/protocol/ImPacketCodec.kt
package com.buyansong.im.protocol

import java.nio.ByteBuffer
import java.nio.ByteOrder

object ImPacketCodec {
    // B6 自定义二进制协议帧头：
    // magic(2) + version(1) + length(4) + cmd(4) + body(N) + crc(4)
    // 这里的 magic/version 也是 Wireshark 里最容易先定位的标记。
    const val MAGIC: Short = 0xCAFE.toShort()
    const val VERSION: Byte = 1

    private const val HEADER_SIZE = 11
    private const val CRC_SIZE = 4

    fun encode(packet: ImPacket): ByteArray {
        val body = packet.body
        // 协议全部字段都用大端序，和网络抓包里看到的字节顺序一致。
        val buffer = ByteBuffer
            .allocate(HEADER_SIZE + body.size + CRC_SIZE)
            .order(ByteOrder.BIG_ENDIAN)

        // Header:
        // 0..1  : magic = CA FE
        // 2     : version = 01
        // 3..6  : body length
        // 7..10 : command id
        buffer.putShort(MAGIC)
        buffer.put(packet.version)
        buffer.putInt(body.size)
        buffer.putInt(packet.cmd)
        buffer.put(body)

        // CRC 覆盖的是“CRC 之前的所有字节”，不包含尾部 4 字节 CRC 自身。
        val withoutCrc = buffer.position()
        buffer.putInt(Crc32.of(buffer.array(), 0, withoutCrc))
        return buffer.array()
    }

    fun decode(bytes: ByteArray): ImPacket {
        if (bytes.size < HEADER_SIZE + CRC_SIZE) {
            throw ProtocolException("Packet too short")
        }

        // 先校验尾部 CRC，再做字段解析，避免解析到损坏包。
        val expectedCrc = ByteBuffer.wrap(bytes, bytes.size - CRC_SIZE, CRC_SIZE)
            .order(ByteOrder.BIG_ENDIAN)
            .int
        val actualCrc = Crc32.of(bytes, 0, bytes.size - CRC_SIZE)
        if (expectedCrc != actualCrc) {
            throw ProtocolException("Invalid CRC")
        }

        val buffer = ByteBuffer.wrap(bytes).order(ByteOrder.BIG_ENDIAN)
        val magic = buffer.short
        if (magic != MAGIC) {
            throw ProtocolException("Invalid magic")
        }

        val version = buffer.get()
        if (version != VERSION) {
            throw ProtocolException("Unsupported version")
        }

        val length = buffer.int
        if (length < 0 || bytes.size != HEADER_SIZE + length + CRC_SIZE) {
            throw ProtocolException("Invalid body length")
        }

        val cmd = buffer.int
        val body = ByteArray(length)
        buffer.get(body)
        return ImPacket(cmd = cmd, body = body, version = version)
    }
}
